File: Creational_Patterns/Java/Prototype.java
import java.util.HashMap;
import java.util.Map;

// Vehicle is the base Prototype
interface Vehicle {

	void print();

	Vehicle clone();

}

// Motorcycle is a concrete Prototype
class Motorcycle implements Vehicle {

	private String model;

	public Motorcycle(String model) {
		this.model = model;
	}

	@Override
	public void print() {
		System.out.println("Motorcycle");
		System.out.println("Model: " + this.model + "\n");
	}

	@Override
	public Vehicle clone() {
		return new Motorcycle(this.model);
	}

}

// Bus is a concrete Prototype
class Bus implements Vehicle {

	private String model;

	public Bus(String model) {
		this.model = model;
	}

	@Override
	public void print() {
		System.out.println("Bus");
		System.out.println("Model: " + this.model + "\n");
	}

	@Override
	public Vehicle clone() {
		return new Bus(this.model);
	}

}

// Car is a concrete Prototype
class Car implements Vehicle {

	private String model;

	public Car(String model) {
		this.model = model;
	}

	@Override
	public void print() {
		System.out.println("Car");
		System.out.println("Model: " + this.model + "\n");
	}

	@Override
	public Vehicle clone() {
		return new Car(this.model);
	}

}

enum VehicleType
{
	MOTORCYCLE,
	BUS,
	CAR
};

// VehicleFactory is the client
class VehicleFactory {

	private Map<VehicleType, Vehicle> vehicles;

	public VehicleFactory() {
		this.vehicles = new HashMap<VehicleType, Vehicle>();
		this.vehicles.put(VehicleType.MOTORCYCLE, new Motorcycle("Harley-Davidson"));
		this.vehicles.put(VehicleType.BUS, new Bus("Mercedes-Benz"));
		this.vehicles.put(VehicleType.CAR, new Car("Ferrari"));
	}

	public Vehicle createVehicle(VehicleType vehicleType) {
		return vehicles.get(vehicleType).clone();
	}

}

class Main {

	public static void main(String[] args) {
		System.out.println(":: Prototype ::\n");

		VehicleFactory factory = new VehicleFactory();

		Vehicle vehicle = factory.createVehicle(VehicleType.MOTORCYCLE);
		vehicle.print();

		vehicle = factory.createVehicle(VehicleType.BUS);
		vehicle.print();

		vehicle = factory.createVehicle(VehicleType.CAR);
		vehicle.print();
	}

}

File: Creational_Patterns/C++/Prototype.cpp
#include <iostream>
#include <string>
#include <memory>
#include <map>

// Vehicle is the base Prototype
class Vehicle
{
public:
	virtual ~Vehicle() {};
	virtual void print() = 0;
	virtual std::unique_ptr<Vehicle> clone() = 0;
};

// Motorcycle is a concrete Prototype
class Motorcycle : public Vehicle
{
public:
	Motorcycle(std::string model) : m_Model(model) {}

	void print() override
	{
		std::cout << "Motorcycle" << std::endl << "Model: " << m_Model << "\n\n";
	}

	std::unique_ptr<Vehicle> clone() override
	{
		return std::make_unique<Motorcycle>(*this);
	}

private:
	std::string m_Model;
};

// Bus is a concrete Prototype
class Bus : public Vehicle
{
public:
	Bus(std::string model) : m_Model(model) {}

	void print() override
	{
		std::cout << "Bus" << std::endl << "Model: " << m_Model << "\n\n";
	}

	std::unique_ptr<Vehicle> clone() override
	{
		return std::make_unique<Bus>(*this);
	}

private:
	std::string m_Model;
};

// Car is a concrete Prototype
class Car : public Vehicle
{
public:
	Car(std::string model) : m_Model(model) {}

	void print() override
	{
		std::cout << "Car" << std::endl << "Model: " << m_Model << "\n\n";
	}

	std::unique_ptr<Vehicle> clone() override
	{
		return std::make_unique<Car>(*this);
	}

private:
	std::string m_Model;
};

enum VehicleType
{
	MOTORCYCLE,
	BUS,
	CAR
};

// VehicleFactory is the client
class VehicleFactory
{
public:
	VehicleFactory()
	{
		m_Vehicles[MOTORCYCLE] = std::make_unique<Motorcycle>("Harley-Davidson");
		m_Vehicles[BUS] = std::make_unique<Bus>("Mercedes-Benz");
		m_Vehicles[CAR] = std::make_unique<Car>("Ferrari");
	}

	std::unique_ptr<Vehicle> createVehicle(VehicleType vehicleType)
	{
		return m_Vehicles[vehicleType]->clone();
	}

private:
	std::map<VehicleType, std::unique_ptr<Vehicle>> m_Vehicles;
};

int main()
{
	std::cout << ":: Prototype ::\n" << std::endl;

	VehicleFactory factory;

	auto vehicle = factory.createVehicle(MOTORCYCLE);
	vehicle->print();

	vehicle = factory.createVehicle(BUS);
	vehicle->print();

	vehicle = factory.createVehicle(CAR);
	vehicle->print();

	system("PAUSE");

	return 0;
}
